src/Microsoft.AspNetCore.Razor.Language/StreamSourceDocument.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class StreamSourceDocument : RazorSourceDocument
+    {
+        // Internal for testing
+        internal readonly RazorSourceDocument _innerSourceDocument;
+
+        private readonly byte[] _streamBytes;
+        private byte[] _checksum;
+
+        public StreamSourceDocument(Stream stream, Encoding encoding, string fileName)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _streamBytes = ReadStreamBytes(stream);
+            _innerSourceDocument = CreateInnerSourceDocument(stream, encoding, fileName);
```

Comment:
> Was trying to be lazy about the checksum calculation; with how the world currently works we always evaluate it anyways so I agree with you.

Going to leave the `StringSourceDocument` and `LargeTextSourceDocument` lazy since they're wrapped by `StreamSourceDocument` (their checksums are never calculated).